clang/test/SemaHLSL/BuiltIns/and-errors.hlsl

```diff
@@ -0,0 +1,27 @@
+// RUN: %clang_cc1 -finclude-default-header -triple \
+// RUN:   dxil-pc-shadermodel6.3-library %s \
+// RUN:   -emit-llvm -O1 -verify
+
+bool test_too_few_arg(bool a) {
+  return __builtin_hlsl_and(a);
+  // expected-error@-1 {{too few arguments to function call, expected 2, have 1}}
+}
+
+bool test_too_many_arg(bool a) {
+  return __builtin_hlsl_and(a, a, a);
+  // expected-error@-1 {{too many arguments to function call, expected 2, have 3}}
+}
+
+bool2 test_mismatched_args(bool2 a, bool3 b) {
+  return __builtin_hlsl_and(a, b);
+  // expected-error@-1 {{all arguments to '__builtin_hlsl_and' must have the same type}}
+}
+
+struct S {
+  bool a;
+};
+
+bool test_invalid_type_conversion(S s) {
+  return __builtin_hlsl_and(s, s);
```

Comment:
This call should be failing because `s` isn't a bool, not because you can't cast `S` back to bool in the return type.

Reply:
Actually this is a good catch. The error message is worded strangely-- it implies that `__builtin_hlsl_and` is trying to return a struct of type `S` and is failing because `S` can't be converted to a bool.

I will fix this by adding a check to SemaHLSL that the argument type has a bool representation, and emitting a better error message.

Reply:
Though I thought there would be something to catch errors like this when the argument type doesn't match any of the ones declared in `hlsl_intrinsics.h`

Reply:
Actually, there is something like that, but it seems `__builtin_hlsl_and` is triggering something else.
The `D3DCOLORtoUBYTE` function doesn't have any SemaHLSL and [yet has a similar test which emits a more accurate error message](https://github.com/llvm/llvm-project/blob/main/clang/test/SemaHLSL/BuiltIns/D3DCOLORtoUBYTE4-errors.hlsl#L28). Perhaps this is just because `D3DCOLORtoUBYTE` is a header-only implementation?

Reply:
I understand now why `__builtin_hlsl_and` was trying to return a struct of type `S`: it was specified in `SemaHLSL.cpp` that the return type is equal to the type of the first argument. Hence, when `S` is passed in, `__builtin_hlsl_and`'s return type is also `S` and therefore the error message now makes sense because `S` can not be converted to a bool to satisfy the return type of the `and` function.

Hence the need to add a check to ensure that the types of the arguments passed to `__builtin_hlsl_and` are bools.